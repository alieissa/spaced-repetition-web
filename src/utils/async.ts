/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
