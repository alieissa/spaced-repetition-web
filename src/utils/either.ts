/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
