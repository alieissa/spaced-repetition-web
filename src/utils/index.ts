/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
