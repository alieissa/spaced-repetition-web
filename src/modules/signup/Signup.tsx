/** @format */

import { FormikState, useFormik } from 'formik'
import { ChangeEvent, useEffect } from 'react'
import { Container, Form, Loader, Message } from 'semantic-ui-react'
import { SPButton, SPHeader, SPInput, SPText } from 'src/components'
import { styles } from 'src/styles'
import * as Yup from 'yup'
import { useSignupMutation } from './signup.hooks'

type UserForm = {
  firstName: string
  lastName: string
  email: string
  password: string
  confirmedPassword: string
}
const UserFormSchema = Yup.object().shape({
  firstName: Yup.string(),
  lastName: Yup.string(),
  email: Yup.string().email('Invalid email').required('Required'),
  password: Yup.string().required('Required'),
  confirmedPassword: Yup.string().required('Required'),
})
type Props = {
  form: FormikState<UserForm>
  onChangeFirstName: (e: ChangeEvent<HTMLInputElement>) => void
  onChangeLastName: (e: ChangeEvent<HTMLInputElement>) => void
  onChangeEmail: (e: ChangeEvent<HTMLInputElement>) => void
  onChangePassword: (e: ChangeEvent<HTMLInputElement>) => void
  onChangeConfirmedPassword: (e: ChangeEvent<HTMLInputElement>) => void
  onSignup: VoidFunction
}
const SignupComponent = (props: Props) => {
  const emailError = props.form.touched.email && !!props.form.errors.email
  const passwordError =
    props.form.touched.password && !!props.form.errors.password
  const confirmedPasswordError =
    props.form.touched.confirmedPassword &&
    !!props.form.errors.confirmedPassword

  return (
    <Form
      onSubmit={props.onSignup}
      className="flex-column bordered p-1r"
      name="signup-form"
    >
      <Form.Field>
        <label htmlFor="first name">First name</label>
        <SPInput
          data-testid="signup-form-first-name-input"
          title="first name"
          disabled={props.form.isSubmitting}
          placeholder="First name"
          value={props.form.values.firstName}
          onChange={props.onChangeFirstName}
        />
      </Form.Field>
      <Form.Field>
        <label>Last name</label>
        <SPInput
          data-testid="signup-form-last-name-input"
          title="last name"
          disabled={props.form.isSubmitting}
          placeholder="Last name"
          value={props.form.values.lastName}
          onChange={props.onChangeLastName}
        />
      </Form.Field>
      <Form.Field error={emailError}>
        <label>Email</label>
        <SPInput
          data-testid="signup-form-email-input"
          name="email"
          title="email"
          disabled={props.form.isSubmitting}
          placeholder="Email"
          value={props.form.values.email}
          onChange={props.onChangeEmail}
        />
        {emailError && (
          <SPText
            data-testid="signup-form-email-error"
            style={styles['mt-0.5rem']}
          >
            {props.form.errors.email}
          </SPText>
        )}
      </Form.Field>
      <Form.Field error={passwordError}>
        <label>Password</label>
        <SPInput
          data-testid="signup-form-password-input"
          title="password"
          disabled={props.form.isSubmitting}
          placeholder="Password"
          type="password"
          value={props.form.values.password}
          onChange={props.onChangePassword}
        />
        {passwordError && (
          <SPText
            data-testid="signup-form-password-error"
            style={styles['mt-0.5rem']}
          >
            {props.form.errors.password}
          </SPText>
        )}
      </Form.Field>
      <Form.Field error={confirmedPasswordError}>
        <label>Confirm password</label>
        <SPInput
          data-testid="signup-form-confirmed-password-input"
          title="confirm password"
          disabled={props.form.isSubmitting}
          placeholder="Confirm password"
          type="password"
          value={props.form.values.confirmedPassword}
          onChange={props.onChangeConfirmedPassword}
        />
        {confirmedPasswordError && (
          <SPText
            data-testid="signup-form-confirmed-password-error"
            style={styles['mt-0.5rem']}
          >
            {props.form.errors.confirmedPassword}
          </SPText>
        )}
      </Form.Field>
      <SPButton
        type="submit"
        disabled={props.form.isSubmitting}
        className="align-self-end"
      >
        Submit
      </SPButton>
    </Form>
  )
}

function Signup() {
  const signupMutation = useSignupMutation()

  const form = useFormik({
    initialValues: {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmedPassword: '',
    },
    validationSchema: UserFormSchema,
    onSubmit: (data: any) => signupMutation.mutate(data),
  })

  useEffect(() => {
    const isSubmitting = signupMutation.status === 'loading'
    form.setSubmitting(isSubmitting)
  }, [form, signupMutation.status])

  const handleChangeFirstName = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('firstName', e.target.value)
  }
  const handleChangeLastName = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('lastName', e.target.value)
  }
  const handleChangeEmail = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('email', e.target.value)
  }
  const handleChangePassword = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('password', e.target.value)
  }
  const handleChangeConfirmedPassword = (e: ChangeEvent<HTMLInputElement>) => {
    form.setFieldValue('confirmedPassword', e.target.value)
  }

  const renderSignupComponent = () => {
    const commonProps = {
      form,
      onChangeFirstName: handleChangeFirstName,
      onChangeLastName: handleChangeLastName,
      onChangeEmail: handleChangeEmail,
      onChangePassword: handleChangePassword,
      onChangeConfirmedPassword: handleChangeConfirmedPassword,
      onSignup: form.submitForm,
    }
    switch (signupMutation.status) {
      case 'idle': {
        return (
          <div data-testid="signup-form">
            <SignupComponent {...commonProps} />
          </div>
        )
      }
      case 'loading': {
        return (
          <>
            <Loader active />
            <SignupComponent {...commonProps} />
          </>
        )
      }
      case 'success': {
        return (
          <Message positive>
            <Message.Header>Signup successful</Message.Header>
            <p>An email has been sent to you.</p>
          </Message>
        )
      }
      case 'error': {
        return (
          <>
            <Message negative>
              <Message.Header>Signup failed</Message.Header>
              <p>Please try again.</p>
            </Message>
            <SignupComponent {...commonProps} />
          </>
        )
      }
    }
  }

  return (
    <Container>
      <div className="justify-space-between bordered p-1r mb-1r">
        <SPHeader as="h1">Spaced Reps</SPHeader>
      </div>
      {renderSignupComponent()}
    </Container>
  )
}

export default Signup
