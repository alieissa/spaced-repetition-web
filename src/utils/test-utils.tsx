/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
