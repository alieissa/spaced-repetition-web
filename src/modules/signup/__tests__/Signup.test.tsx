/** @format */

import '@testing-library/jest-dom'
import {
  act,
  fireEvent,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { renderWithProviders } from 'src/utils/test-utils'
import Signup from '../Signup'

const signupUrl = `${process.env.REACT_APP_API_ENDPOINT}/users/register`

export const handlers = [
  rest.post(signupUrl, (__, res, ctx) => {
    return res(
      ctx.json({ email: 'jonhnybravo1@gmail.com', password: 'heythere!' }),
    )
  }),
]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('Signup', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<Signup />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    test('display initial', async () => {
      renderWithProviders(<Signup />)
      expect(screen.getByTestId('signup-form')).toBeInTheDocument()
    })

    test('display loading', async () => {
      const user = userEvent.setup()

      renderWithProviders(<Signup />)

      const emailInput = screen.getByRole('textbox', { name: 'email' })
      const passwordInput = screen.getByTitle('password')
      await act(() => user.type(emailInput, 'johnnyb1@gmail.com'))
      await act(() => user.type(passwordInput, 'heythere!'))
      fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

      const testId = 'signup-form-loading'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
      // See https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b
      await waitForElementToBeRemoved(() => screen.queryByTestId(testId))
    })

    test('display failure', async () => {
      server.use(
        rest.post(signupUrl, (__, res, ctx) =>
          res(ctx.json({ email: 'dfsfsdf' }), ctx.status(422)),
        ),
      )

      renderWithProviders(<Signup />)

      const user = userEvent.setup()
      const emailInput = screen.getByRole('textbox', { name: 'email' })
      const passwordInput = screen.getByTitle('password')
      await act(() => user.type(emailInput, 'johnnyb1@gmail.com'))
      await act(() => user.type(passwordInput, 'heythere!'))

      fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

      const testId = 'signup-form-error'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })

    test('display success', async () => {
      renderWithProviders(<Signup />)

      const user = userEvent.setup()
      const emailInput = screen.getByRole('textbox', { name: 'email' })
      const passwordInput = screen.getByTitle('password')
      await act(() => user.type(emailInput, 'johnnyb1@gmail.com'))
      await act(() => user.type(passwordInput, 'heythere!'))

      fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

      const testId = 'signup-form-successful'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })
  })
})
