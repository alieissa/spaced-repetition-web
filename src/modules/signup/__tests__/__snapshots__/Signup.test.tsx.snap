// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup view should render correctly 1`] = `
<DocumentFragment>
  <div
    class="ui segment"
    data-testid="signup-form"
  >
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <label>
          Email
        </label>
        <input
          placeholder="Email"
          title="email"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label>
          Password
        </label>
        <input
          placeholder="Password"
          title="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="ui button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
